rfc/003-skip-values.md

```diff
@@ -0,0 +1,157 @@
+
+## Proposal
+
+This proposal aims to enhance the Record Linker Algorithm configuration by introducing a
+new `skip_values` section, allowing users to specify conditions under which certain field
+values should be excluded from comparisons. This approach is beneficial because:
+- Algorithm configuration is already user-defined, making the addition of a
+    `skip_values` section a natural extension.
+- Organizations may need to manage skip conditions over time as their data evolves,
+    and this system provides the flexibility to accommodate those changes.
+
+The `skip_values` section will be defined at the algorithm level (not per pass) and will
+contain a list of conditions. Each condition will have two elements: `feature` and `values`.
+The `feature` specifies the field to which the condition applies, while `values` indicates
+the case-insensitive list of values to skip. The `*` wildcard can be used to apply in the
+`feature` field to apply the condition to all fields. Example:
+
+```json
+...
+"skip_values": [
+  {
+    "feature": "NAME",
+    "values": ["John Doe", "Jane Doe"]
+  },
+  {
+    "feature": "IDENTIFIER:SS",
+    "values": ["999-99-9999"]
+  },
+  {
+    "feature": "*",
+    "values": ["unknown", "not specified"]
+  }
+]
+```
+
+In this example, the algorithm will exclude fields from comparison if they match:
+- `John Doe` or `Jane Doe` in the patient's name
+- `999-99-9999` in the patient's social security number
+- `unknown` or `not specified` in any patient field
+
+### Pre-Processing Details
+
+Fields that match skip conditions will be removed only from the data copy used for
+evaluation, not from the original data stored in the MPI. This ensures that users can
+modify skip rules over time or re-run the algorithm without risking data loss.
+
+## Alternatives Considered
+
+Two alternatives were considered, but both have significant drawbacks. The first option is
+to apply a data normalization step upon receiving the data to remove any placeholder values.
+While similar to the proposed approach, the key difference is that the data is permanently
+altered, with placeholder values removed before being persisted. This poses a problem if
+users later update their skip values, as any removed data would no longer be available in
+the MPI for future use.
+
+The second alternative is to apply specific rules during evaluation to determine whether a
+value should be used in a comparison or ignored as a placeholder. The main flaw with this
+approach is the lack of context. For example, if comparing the `FIRST_NAME` field and the
+incoming record has the value "John", this is typically a valid name for comparison.
+However, if the `LAST_NAME` is "Doe", it may be recognized as a placeholder. Performing the
+cleaning upfront allows for a comprehensive review of all data fields before evaluation,
+ensuring that additional context is considered.
```

Comment:
I'm not sure I follow the last sentence here. I definitely get the idea that we want to apply skip values before evaluation, but where is this exactly sitting? In my mind it's the `feature_iter` yield function (i.e. for each of `FIRST_NAME` and `LAST_NAME`, if `FIRST_NAME == John and LAST_NAME == Doe yield ''`), but is that what you mean here?

Reply:
The cleaning step as indicated in step 3 of the Implementation Plan, will likely need to duplicate some of the logic in `feature_iter`. It will accept a record object and set of cleaning rules, then return a new record object sans any features matched in the cleaning rules.

It's this new copy, returned from the cleaning function, that will be eventually passed to the `feature_iter`.

Reply:
Okay, so this is where we'd do the thing I mentioned in my comment above (comparing John Doe to John Smith)? I.e. the cleaning function would see that one of the records is a John Doe and then "blank" out the name values, before `feature_iter` ever gets ahold of it?

Reply:
If that's the case, then will we need to apply the cleaning function to blocked candidates returned from the MPI? In other words, if we've previously processed some John Does, and they get created as new Person Clusters in the MPI, then anytime in the future we get a record who might block into John Doe, we'll need to make sure we don't do a comparison on the fetched candidates. Although, writing that out, there might be some trickier rules around blocking with John Doe anyway.

Reply:
You know, my original thought was we only need to "clean" the incoming record.  As long as we clean one of the two objects being compared, we'd be safe.  But if the incoming record is "John Smith" and the blocked candidate is "John Doe" then we'd get a perfect match on FIRST_NAME, when we shouldn't!

You're 💯 right, we need to clean both the incoming record AND the blocked candidates.